### PR TITLE
fixes case sensitivity for SameSite directive

### PR DESCRIPTION
**Pull request by wilcol**
This fix handles case sensitivity for SameSite directive in cookie headers.

Prior to this fix the SameSite values were case sensitive compared with their capitalized names (`Lax`, `Strict`, `None`). This fix allows for case insensitive comparison, allowing values like `lax` and `LAX`. 

See https://github.com/zendframework/zend-http/issues/206 for more information

### CODE REVIEW

File: src/Header/SetCookie.php [hunk 1]
```diff
@@ -337,7 +337,8 @@ public function getFieldValue()
         }
 
         $sameSite = $this->getSameSite();
-        if ($sameSite !== null && in_array($sameSite, self::SAME_SITE_ALLOWED_VALUES, true)) {
+        if ($sameSite !== null
+            && in_array(strtolower($sameSite), array_map('strtolower', self::SAME_SITE_ALLOWED_VALUES), true)) {
```
Comment (michalbundyra) [inline src/Header/SetCookie.php]: I would suggest to change the values in the constant `SAME_SITE_ALLOWED_VALUES` as it is marked as `internal`. Then we don't need to lowercase it here.

File: src/Header/SetCookie.php [hunk 2]
```diff
@@ -251,15 +251,15 @@ public function __construct(
...
-             ->setSameSite($sameSite);
+            ->setValue($value)
+            ->setVersion($version)
+            ->setMaxAge($maxAge)
+            ->setDomain($domain)
+            ->setExpires($expires)
+            ->setPath($path)
+            ->setSecure($secure)
+            ->setHttpOnly($httponly)
+            ->setSameSite($sameSite);
```
Comment (michalbundyra) [inline src/Header/SetCookie.php]: Please revert, the change is unrelated to the hotfix.

File: src/Header/SetCookie.php [hunk 3]
```diff
@@ -618,7 +619,8 @@ public function getSameSite()
      */
     public function setSameSite($sameSite)
     {
-        if ($sameSite !== null && ! in_array($sameSite, self::SAME_SITE_ALLOWED_VALUES, true)) {
+        if ($sameSite !== null
+            && ! in_array(strtolower($sameSite), array_map('strtolower', self::SAME_SITE_ALLOWED_VALUES), true)) {
```
Comment (michalbundyra) [inline src/Header/SetCookie.php]: As above - if we change values in the constant then we don't need to have array_map here.

File: test/Header/SetCookieTest.php [hunk 4]
```diff
@@ -161,6 +187,30 @@ public function testSetCookieFromStringCanCreateSingleHeader()
         $this->assertTrue($setCookieHeader->isSecure());
         $this->assertTrue($setCookieHeader->isHttponly());
         $this->assertEquals(setCookie::SAME_SITE_STRICT, $setCookieHeader->getSameSite());
+        $this->assertEquals(
+            'myname=myvalue; Expires=Wed, 13-Jan-2021 22:23:01 GMT; Domain=docs.foo.com; '
+            . 'Path=/accounts; Secure; HttpOnly; SameSite=Strict',
+            $setCookieHeader->getFieldValue()
+        );
+
+        $setCookieHeader = SetCookie::fromString(
```
Comment (michalbundyra) [inline test/Header/SetCookieTest.php]: It should be separate test case.
I see it is like that in that test case, but it's wrong...

File: CHANGELOG.md [hunk 5]
```diff
@@ -44,7 +44,7 @@ All notable changes to this project will be documented in this file, in reverse
 
 ### Fixed
 
-- Nothing.
+- fixes case sensitivity for SameSite directive.
```
Comment (michalbundyra) [inline CHANGELOG.md]: Can be fixed on merge: it should contain link to the PR as we have for other changelogs

**Review by michalbundyra — CHANGES_REQUESTED** [review]
@wilcol Thanks for your contribution. I have couple minor comments. In general it looks good. See my comments below. Thanks!

**Comment by wilcol** [comment]
@michalbundyra, I processed your feedback. Thank you.

File: src/Header/SetCookie.php [hunk 6]
```diff
@@ -39,9 +39,9 @@ class SetCookie implements MultipleHeaderInterface
      * @internal
      */
     const SAME_SITE_ALLOWED_VALUES = [
-        self::SAME_SITE_STRICT,
-        self::SAME_SITE_LAX,
-        self::SAME_SITE_NONE,
+        'strict',
+        'lax',
+        'none',
```
Comment (Xerkus) [inline src/Header/SetCookie.php]: I would suggest it to be changed to
```php
 const SAME_SITE_ALLOWED_VALUES = [
        'strict' = self::SAME_SITE_STRICT,
        'lax' = self::SAME_SITE_LAX,
        'none' = self::SAME_SITE_NONE,
    ];
```
Lowercase assertion can be made against key and additionally that will allow to set canonical form value:
```
$this->sameSite = self::SAME_SITE_ALLOWED_VALUES[strtolower($sameSite)];
```

Comment (Xerkus) [reply src/Header/SetCookie.php]: Considering that users might check value against constant, please make that change and add additional test ensuring case insensitive value is set to canonical form.

**Comment by wilcol** [comment]
@Xerkus, I processed your feedback.

File: src/Header/SetCookie.php [hunk 7]
```diff
@@ -618,13 +618,18 @@ public function getSameSite()
      */
     public function setSameSite($sameSite)
     {
-        if ($sameSite !== null && ! in_array($sameSite, self::SAME_SITE_ALLOWED_VALUES, true)) {
-            throw new Exception\InvalidArgumentException(sprintf(
-                'Invalid value provided for SameSite directive: "%s"; expected one of: Strict, Lax or None',
-                is_scalar($sameSite) ? $sameSite : gettype($sameSite)
-            ));
+        if ($sameSite !== null) {
```
Comment (Xerkus) [inline src/Header/SetCookie.php]: Early return trick can be used to increase readability by decreasing indentation level and avoiding else statements:
```php
if ($sameSite === null) {
    $this->sameSite = null;
    return;
}
```

**Review by Xerkus — CHANGES_REQUESTED** [review]

**Comment by Xerkus** [comment]
Please also add a separate test for asserting that SameSite value is canonicalized in setSameSite

**Comment by wilcol** [comment]
@Xerkus all done!

File: src/Header/SetCookie.php [hunk 8]
```diff
@@ -337,7 +337,7 @@ public function getFieldValue()
         }
 
         $sameSite = $this->getSameSite();
-        if ($sameSite !== null && in_array($sameSite, self::SAME_SITE_ALLOWED_VALUES, true)) {
+        if ($sameSite !== null && array_key_exists(strtolower($sameSite), self::SAME_SITE_ALLOWED_VALUES)) {
```
Comment (Xerkus) [inline src/Header/SetCookie.php]: coding style: used functions must be imported with `use function` statements (under `use` block, separated by blank line)
Import just those you touched, that will be enough.

File: test/Header/SetCookieTest.php [hunk 9]
```diff
@@ -105,6 +131,51 @@ public function testSetInvalidSameSiteDirectiveValueViaSetter()
...
+            'Wed, 13-Jan-2021 22:23:01 GMT',
+            '/accounts',
+            'docs.foo.com',
+            true,
+            true,
+            99,
+            9,
+            SetCookie::SAME_SITE_STRICT
+        );
+        $this->assertEquals(SetCookie::SAME_SITE_STRICT, $setCookieHeader->getSameSite());
```
Comment (Xerkus) [inline test/Header/SetCookieTest.php]: Please also test `setSameSite()` explicitly. Just once would be enough.

**Review by Xerkus — CHANGES_REQUESTED** [review]
Looks  good, just a bit of extra changes.

**Review by Xerkus — APPROVED** [review]
Perfect, thanks!

**Comment by weierophinney** [comment]
As this was a bugfix, I rebased against master and merged there for a new 2.11.2 release.

Thanks, @wilcol !
